libraries/bot-builder/src/main/java/com/microsoft/bot/builder/CloudAdapterBase.java

```diff
@@ -0,0 +1,456 @@
+import org.slf4j.LoggerFactory;
+
+import java.net.HttpURLConnection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An adapter that implements the Bot Framework Protocol and can be hosted in different cloud environments
+ * both public and private.
+ */
+public abstract class CloudAdapterBase extends BotAdapter {
+
+    protected BotFrameworkAuthentication botFrameworkAuthentication;
+    protected Logger logger = LoggerFactory.getLogger(CloudAdapterBase.class);
+
+    /**
+     * Gets a {@link Logger} to use within this adapter and its subclasses.
+     * @return The {@link Logger} instance for this adapter.
+     */
+    public Logger getLogger() {
+        return logger;
+    }
+
+    /**
+     * Gets the {@link BotFrameworkAuthentication} instance for this adapter.
+     * @return The {@link BotFrameworkAuthentication} instance for this adapter.
+     */
+    public BotFrameworkAuthentication getBotFrameworkAuthentication() {
+        return botFrameworkAuthentication;
+    }
+
+    /**
+     * Initializes a new instance of the {@link CloudAdapterBase} class.
+     * @param withBotFrameworkAuthentication The cloud environment used for validating and creating tokens.
+     */
+    protected CloudAdapterBase(BotFrameworkAuthentication withBotFrameworkAuthentication) {
+        if (withBotFrameworkAuthentication == null) {
+            throw new IllegalArgumentException("withBotFrameworkAuthentication cannot be null");
+        }
+        this.botFrameworkAuthentication = withBotFrameworkAuthentication;
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public CompletableFuture<ResourceResponse[]> sendActivities(TurnContext context, List<Activity> activities) {
+        if (context == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("context"));
+        }
+
+        if (activities == null) {
+            return Async.completeExceptionally(new IllegalArgumentException("activities"));
+        }
+
+        if (activities.size() == 0) {
+            return Async.completeExceptionally(
+                new IllegalArgumentException("Expecting one or more activities, but the array was empty.")
+            );
+        }
+
+        logger.info(String.format("sendActivities for %d activities.", activities.size()));
+
+        return CompletableFuture.supplyAsync(() -> {
+            ResourceResponse[] responses = new ResourceResponse[activities.size()];
+
+            for (int index = 0; index < activities.size(); index++) {
+                Activity activity = activities.get(index);
+
+                activity.setId(null);
+                ResourceResponse response;
+
+                logger.info(String.format("Sending activity. ReplyToId: %s", activity.getReplyToId()));
+
+                if (activity.isType(ActivityTypes.DELAY)) {
+                    int delayMs = (int) activity.getValue();
+                    try {
```

Comment:
this try/catch is not in C#. Guess it's necessary for the Thread.sleep

Reply:
Exactly